lta/nersc_verifier.py

```diff
@@ -0,0 +1,240 @@
+        self.tape_bath_path = config["TAPE_BASE_PATH"]
+        self.work_retries = int(config["WORK_RETRIES"])
+        self.work_timeout_seconds = float(config["WORK_TIMEOUT_SECONDS"])
+
+    def _do_status(self) -> Dict[str, Any]:
+        """NerscVerifier has no additional status to contribute."""
+        return {}
+
+    def _expected_config(self) -> Dict[str, Optional[str]]:
+        """NerscVerifier provides our expected configuration dictionary."""
+        return EXPECTED_CONFIG
+
+    async def _do_work(self) -> None:
+        """Perform a work cycle for this component."""
+        self.logger.info("Starting work on Bundles.")
+        work_claimed = True
+        while work_claimed:
+            work_claimed = await self._do_work_claim()
+        self.logger.info("Ending work on Bundles.")
+
+    async def _do_work_claim(self) -> bool:
+        """Claim a bundle and perform work on it."""
+        # 1. Ask the LTA DB for the next Bundle to be verified
+        # configure a RestClient to talk to the LTA DB
+        lta_rc = RestClient(self.lta_rest_url,
+                            token=self.lta_rest_token,
+                            timeout=self.work_timeout_seconds,
+                            retries=self.work_retries)
+        self.logger.info("Asking the LTA DB for a Bundle to verify at NERSC with HPSS.")
+        pop_body = {
+            "claimant": f"{self.name}-{self.instance_uuid}"
+        }
+        response = await lta_rc.request('POST', '/Bundles/actions/pop?dest=NERSC&status=verifying', pop_body)
+        self.logger.info(f"LTA DB responded with: {response}")
+        bundle = response["bundle"]
+        if not bundle:
+            self.logger.info("LTA DB did not provide a Bundle to verify at NERSC with HPSS. Going on vacation.")
+            return False
+        # process the Bundle that we were given
+        if await self._verify_bundle_in_hpss(lta_rc, bundle):
+            await self._add_bundle_to_file_catalog(bundle)
+            await self._update_bundle_in_lta_db(lta_rc, bundle)
+        return True
+
+    async def _add_bundle_to_file_catalog(self, bundle: BundleType) -> bool:
+        """Add a FileCatalog entry for the bundle, then update existing records."""
+        # configure a RestClient to talk to the File Catalog
+        fc_rc = RestClient(self.file_catalog_rest_url,
+                           token=self.file_catalog_rest_token,
+                           timeout=self.work_timeout_seconds,
+                           retries=self.work_retries)
+        # determine the path where the bundle is stored on hpss
+        data_warehouse_path = bundle["path"]
+        basename = os.path.basename(bundle["bundle_path"])
+        stupid_python_path = os.path.sep.join([self.tape_bath_path, data_warehouse_path, basename])
+        hpss_path = os.path.normpath(stupid_python_path)
+        # create a File Catalog entry for the bundle itself
+        file_record = {
+            "uuid": bundle["uuid"],
```

Comment:
Hopefully there is not an overlap; that would seem to require a UUID collision, which ought to be extremely rare.

I re-used the UUID here because it seemed weird to have a `logical_name` like `/home/projects/icecube/data/exp/IceCube/2019/filtered/PFFilt/0304/9203b4e0-0cf7-4704-b332-80d15db4d3e5.zip` with a `uuid` field that wasn't `9203b4e0-0cf7-4704-b332-80d15db4d3e5`.

There is something to be said for orthogonal ID spaces; that perhaps allowing LTA to call something by a UUID, and then also have the FC call something else by the same UUID is like playing with fire. I could imagine a scenario where all the jsons are merged into one gigantic omni collection and all queries are always omni queries, then we see where reuse of a "unique" identifier was a bad idea.

However, I was a bit emboldened by the idea that the LTA entity is ephemeral and the FC entity is durable. My thought was the LTA record would disappear so the FC could reuse the UUID without much worry and it would match the filename nicely.

That said, it's just a 'nice to have'; if the ID re-use seems like bad mojo, this is an easy fix.

Reply:
I guess as long as we are using uuid1 it should be ok.  That's time-based, so unless we're doing a lot of work really close in time it shouldn't matter.

Also, I guess the FC will reject the request if there's an overlap.